F dplan 12665 adjust phases usage

sierrazamudiodemos opened the pull request:
### Ticket
https://demoseurope.youtrack.cloud/issue/DPLAN-12665/Verfahrensschritte-in-Detailansicht-funktionieren-nicht-richtig

This is a refactor / left over of the ticket above.
- We want to treat phases as objects.
- We want to put in one method, the logic that based on the field "publicStatement" of the stn, the external or internal phases are taken
-- If publicStatement is internal, then internalPhases are taken
-- if publicStatement is external, then externalPhases are taken
- We want to move that logic that I just described above into the BE

In order to do the points above, this PR does the following:
- It has a new StatementPhaseService which does 3 main things:
-- It checks if the phase of the current Statement is a valid one
-- It provides a list of available Phases based on the phase and the publicStatement of the Statement
-- It creates the objects to represent the phases 
-- The method getPhaseName from the StatementService was also adjusted to consume the logic on the StatementPhaseService

### How to review/test
- Trigger api call to check the new fields: phaseStatement in OriginalStn and Stn, 
- Trigger api call to check the new field availablePhases in Stn
- For the availablePhases list: This list returns internal or external Phases based on the "phase" of the stn

### Linked PRs (optional)
Existing PR already merged in integration: [ttps://github.com/demos-europe/demosplan-core/pull/3871/files](https://github.com/demos-europe/demosplan-core/pull/3871/files)

Delete the checkbox if it doesn't apply/isn't necessary.

- [ ] Tests updated/created
- [ ] Update documentation
- [ ] Link all relevant tickets
- [ ] Move the tickets on the board accordingly
- [ ] Data-Cy attributes added/updated ([conventions](https://dplan-documentation.demos-europe.eu/development/guidelines-conventions/coding-styleguides/twig_html.html#guideline-for-naming-cypress-hooks))

breierDemos reviewed (changes requested):
Hey, I think it is a good Idea to improve this. Thank you for that!
I think we could adjust the naming. Instead of PhaseVO we could name it ProcedurePhaseVO, the properties of the resourcetypes procedurePhase and availableProcedurePhases instead of phaseStatement and availablePhases. So it is clear that we are talking about procedure phases. 
What do you think?

breierDemos commented on demosplan/DemosPlanCoreBundle/Logic/Statement/StatementPhaseService.php:
```diff
@@ -0,0 +1,78 @@
+                $phases[] = $this->createPhaseVO($internalPhase, Permissions::PROCEDURE_PERMISSION_SCOPE_EXTERNAL);
+            }
+
+            return $phases;
```
:heavy_plus_sign:

sierrazamudiodemos replied:
done dc52cfb4e89d863bf4853976c0617c5de93be44a

breierDemos commented on demosplan/DemosPlanCoreBundle/Logic/Statement/StatementPhaseService.php:
```diff
@@ -0,0 +1,78 @@
+                $phases[] = $this->createPhaseVO($internalPhase, Permissions::PROCEDURE_PERMISSION_SCOPE_INTERNAL);
+            }
+
+            return $phases;
```
:heavy_plus_sign:

sierrazamudiodemos replied:
done dc52cfb4e89d863bf4853976c0617c5de93be44a

sierrazamudiodemos commented:
> Hey, I think it is a good Idea to improve this. Thank you for that! I think we could adjust the naming. Instead of PhaseVO we could name it ProcedurePhaseVO, the properties of the resourcetypes procedurePhase and availableProcedurePhases instead of phaseStatement and availablePhases. So it is clear that we are talking about procedure phases. What do you think?

@breierDemos done 5f717ab57e582478a9db8cda2984aabd1eb3ec7b good idea ans thanks for the feedback

sierrazamudiodemos commented on demosplan/DemosPlanCoreBundle/ResourceTypes/StatementResourceType.php:
```diff
@@ -428,67 +431,40 @@ static function (Statement $statement, array $newValue): array {
         }
 
-        $configBuilder->phase
+        $configBuilder->phaseStatement
```
Gotcha, I was thinking to remove it already so that the FE can adjust the changes on this PR. Steffi knows about it. What do you thin? @breierDemos

breierDemos replied:
Yes :+1:

breierDemos commented on demosplan/DemosPlanCoreBundle/Logic/Statement/StatementPhaseService.php:
```diff
@@ -0,0 +1,81 @@
+                $phases[] = $this->createProcedurePhaseVO($internalPhase, Permissions::PROCEDURE_PERMISSION_SCOPE_EXTERNAL);
+            }
+
+            return $phases;
```
```suggestion
```

breierDemos commented on demosplan/DemosPlanCoreBundle/Logic/Statement/StatementPhaseService.php:
```diff
@@ -0,0 +1,81 @@
+            foreach ($this->globalConfig->getExternalPhasesAssoc() as $internalPhase) {
+                $phases[] = $this->createProcedurePhaseVO($internalPhase, Permissions::PROCEDURE_PERMISSION_SCOPE_EXTERNAL);
+            }
+
```
```suggestion
```

breierDemos commented on demosplan/DemosPlanCoreBundle/Logic/Statement/StatementPhaseService.php:
```diff
@@ -0,0 +1,81 @@
+            foreach ($this->globalConfig->getInternalPhasesAssoc() as $internalPhase) {
+                $phases[] = $this->createProcedurePhaseVO($internalPhase, Permissions::PROCEDURE_PERMISSION_SCOPE_INTERNAL);
+            }
+
```
```suggestion
```

breierDemos commented on demosplan/DemosPlanCoreBundle/Logic/Statement/StatementPhaseService.php:
```diff
@@ -0,0 +1,81 @@
+                $phases[] = $this->createProcedurePhaseVO($internalPhase, Permissions::PROCEDURE_PERMISSION_SCOPE_INTERNAL);
+            }
+
+            return $phases;
```
```suggestion
```

breierDemos commented on demosplan/DemosPlanCoreBundle/Logic/Statement/StatementPhaseService.php:
```diff
@@ -0,0 +1,81 @@
+            return $phases;
+        }
+
+        return [];
```
```suggestion
        return $phases;
```

breierDemos reviewed (commented):
Lgtm :) I would approve it if the frontend is adjusted.
Some little suggestions:
What do you think about StatementProcedurePhaseResolver instead of StatementPhaseService?

sierrazamudiodemos commented:
@breierDemos I adjusted to validate against isSubmittedCitizen. Let me know what you think. I also removed the left overs of the  permission field_show_internal_procedure_phases_in_dropdown and use overall field_statement_phase

breierDemos commented on demosplan/DemosPlanCoreBundle/Logic/Statement/StatementProcedurePhaseResolver.php:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    /**
+     * // If the publicstatement is internal, then return only the internal phases
```
We should remove the comment or adjust it.

breierDemos commented on demosplan/DemosPlanCoreBundle/ResourceTypes/AbstractStatementResourceType.php:
```diff
@@ -188,10 +188,7 @@ protected function getProperties(): array|ResourceConfigBuilderInterface
-                    $statement->getPhase(),
-                    $statement->getPublicStatement()
-                )
+                fn (Statement $statement): string => $this->statementService->getPhaseName($statement)
```
Wrong parameter. We need string $phaseKey, bool $isSubmittedByCitizen

breierDemos commented on demosplan/DemosPlanCoreBundle/Logic/Statement/StatementService.php:
```diff
@@ -2745,27 +2746,21 @@ public function bulkEditStatementsAddData(StatementBulkEditVO $statementEdit): v
     }
 
-    public function getPhaseName(string $phaseKey, string $publicStatement): string
+    public function getPhaseName(string $phaseKey, bool $isSubmittedByCitizen): string
```
Can we rename this function to getProcedurePhaseName?

sierrazamudiodemos replied:
done

breierDemos commented on demosplan/DemosPlanCoreBundle/Logic/Statement/StatementService.php:
```diff
@@ -2745,27 +2746,21 @@ public function bulkEditStatementsAddData(StatementBulkEditVO $statementEdit): v
      */
     public function getPhaseNameFromArray(array $statement): string
```
Can we rename this function to getProcedurePhaseNameFromArray?

breierDemos commented on demosplan/DemosPlanCoreBundle/ResourceConfigBuilder/OriginalStatementResourceConfigBuilder.php:
```diff
@@ -31,6 +31,7 @@
  * @property-read AttributeConfigBuilderInterface<ClauseFunctionInterface<bool>, Statement> $submitterAndAuthorMetaDataAnonymized
  * @property-read AttributeConfigBuilderInterface<ClauseFunctionInterface<bool>, Statement> $textPassagesAnonymized
  * @property-read AttributeConfigBuilderInterface<ClauseFunctionInterface<bool>, Statement> $textIsTruncated
+ * @property-read AttributeConfigBuilderInterface<ClauseFunctionInterface<bool>,Statement> $phaseStatement
```
Can we rename this property to $procedurePhase?

breierDemos commented on demosplan/DemosPlanCoreBundle/ResourceConfigBuilder/StatementResourceConfigBuilder.php:
```diff
@@ -90,6 +90,8 @@
  * @property-read AttributeConfigBuilderInterface<ClauseFunctionInterface<bool>,Statement> $availableInternalPhases
  * @property-read AttributeConfigBuilderInterface<ClauseFunctionInterface<bool>,Statement> $availableExternalPhases
+ * @property-read AttributeConfigBuilderInterface<ClauseFunctionInterface<bool>,Statement> $phaseStatement
+ * @property-read AttributeConfigBuilderInterface<ClauseFunctionInterface<bool>,Statement> $availablePhases
```
Can we rename this properties to: $procedurePhase and $availableProcedurePhases?

breierDemos reviewed (changes requested):
Hey, thanks for your effort. I have a few comments and name suggestions.

breierDemos reviewed (approved):
Thank you very much, that looks great :smile:

sierrazamudiodemos commented:
huhu @breierDemos fields and methods are renames
huhu @gruenbergerdemos @hwiem 
the originalStatement part needs a small adjustment... instead of phase, procedurePhase should be used. It returns an object (as it in the StatementResourceType) is it possible to adjust it in FE?

gruenbergerdemos commented:
@sierrazamudiodemos Original statement list is adjusted

spiess-demos commented on client/js/components/assessmenttable/DpNewStatement.vue:
```diff
@@ -170,7 +170,7 @@ export default {
     ]),
 
     phases () {
-      if (this.institutionSelected && hasPermission('field_show_internal_procedure_phases_in_dropdown')) {
```
Why can this permission be removed?

sierrazamudiodemos replied:
Because there is an existing permission doing the same: `field_statement_phase`

spiess-demos replied:
but you did not replace the permission but entirely remove it!

spiess-demos replied:
long story short (as i understood it from @sierrazamudiodemos and looking through historic git repos):
- `field_show_internal_procedure_phases_in_dropdown` was introduced to replace a twig override as a refactoring
- this permission however was not properly set in all projects, resulting in it being not effective 
- instead of choosing client-side which phases to display, that logic is now implemented in the backend, so that `availableProcedurePhases` always holds the correct set of phases (internal or external)

spiess-demos reviewed (approved):
lgtm (code review)
